frontend/src/app/pages/exercises/exercise/shared/exercise-map/utility/popup-manager.ts

```diff
@@ -5,22 +5,87 @@ import type {
     ViewContainerRef,
 } from '@angular/core';
 import { isEqual } from 'lodash-es';
-import type { Overlay } from 'ol';
+import type { Feature } from 'ol';
+import { Overlay } from 'ol';
+import type VectorLayer from 'ol/layer/Vector';
+import type VectorSource from 'ol/source/Vector';
 import { Subject, takeUntil } from 'rxjs';
+import type OlMap from 'ol/Map';
 import type { Positioning } from '../../utility/types/positioning';
+import type { FeatureManager } from './feature-manager';
 
 /**
  * A class that manages the creation and destruction of a single popup with freely customizable content
  * that should appear on the {@link popupOverlay}.
  */
 export class PopupManager {
+    /**
+     * If this subject emits options, the specified popup should be toggled.
+     * If it emits undefined, the currently open popup should be closed.
+     */
+    public readonly changePopup$ = new Subject<
+        OpenPopupOptions<any> | undefined
+    >();
+
+    public readonly popupOverlay: Overlay;
     private readonly destroy$ = new Subject<void>();
     private currentlyOpenPopupOptions?: OpenPopupOptions<any>;
+    private popupsEnabled = true;
 
     constructor(
-        private readonly popupOverlay: Overlay,
-        private readonly popoverContent: ViewContainerRef
-    ) {}
+        private readonly popoverContent: ViewContainerRef,
+        private readonly popoverContainer: HTMLDivElement
+    ) {
+        this.popupOverlay = new Overlay({
+            element: this.popoverContainer,
+        });
+    }
+    public setPopupsEnabled(enabled: boolean) {
+        this.popupsEnabled = enabled;
+        if (!enabled) {
+            // Close all open popups
+            this.changePopup$.next(undefined);
+        }
+    }
+
+    public registerPopupTriggers(
+        olMap: OlMap,
+        openLayersContainer: HTMLDivElement,
+        layerFeatureManagerDictionary: Map<
+            VectorLayer<VectorSource>,
+            FeatureManager<any>
+        >
+    ) {
+        olMap.on('singleclick', (event) => {
+            if (!this.popupsEnabled) {
+                return;
+            }
+            olMap.forEachFeatureAtPixel(
+                event.pixel,
+                (feature, layer) => {
+                    // Skip layer when unset
+                    if (layer === null) {
+                        return false;
+                    }
+                    layerFeatureManagerDictionary
+                        .get(layer as VectorLayer<VectorSource>)!
+                        .onFeatureClicked(event, feature as Feature);
+                    // we only want the top one -> a truthy return breaks this loop
+                    return true;
+                },
+                { hitTolerance: 10 }
+            );
+            if (!olMap.hasFeatureAtPixel(event.pixel)) {
+                this.changePopup$.next(undefined);
+            }
```

Comment:
```suggestion
            const hasFeature = olMap.forEachFeatureAtPixel(
                event.pixel,
                (feature, layer) => {
                    // Skip layer when unset
                    if (layer === null) {
                        return false;
                    }
                    layerFeatureManagerDictionary
                        .get(layer as VectorLayer<VectorSource>)!
                        .onFeatureClicked(event, feature as Feature);
                    // we only want the top one -> a truthy return breaks this loop
                    return true;
                },
                { hitTolerance: 10 }
            );
            if (!hasFeature) {
                this.changePopup$.next(undefined);
            }
```

Reply:
hasfeature can be undefined by TypeScript Typings, so I think this introduces additional implicit knowledge

Reply:
Done